src/connector/src/sink/sqlserver.rs

```diff
@@ -0,0 +1,625 @@
+                        return Err(SinkError::SqlServer(anyhow!(format!(
+                            "column {} is primary key in RisingWave sink, but not in SQL Server table {}",
+                            col.name, self.config.table,
+                        ))));
+                    }
+                    if !rw_is_pk && *sql_server_is_pk {
+                        return Err(SinkError::SqlServer(anyhow!(format!(
+                            "column {} is primary key in SQL Server table {}, but not in RisingWave sink",
+                            col.name, self.config.table,
+                        ))));
+                    }
+                }
+            }
+        }
+
+        if !self.is_append_only {
+            let sql_server_pk_count = sql_server_table_metadata
+                .values()
+                .filter(|is_pk| **is_pk)
+                .count();
+            if sql_server_pk_count != self.pk_indices.len() {
+                return Err(SinkError::SqlServer(anyhow!(format!(
+                    "primary key does not match between RisingWave sink ({}) and SQL Server table {} ({})",
+                    self.pk_indices.len(),
+                    self.config.table,
+                    sql_server_pk_count,
+                ))));
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn new_log_sinker(&self, writer_param: SinkWriterParam) -> Result<Self::LogSinker> {
+        Ok(SqlServerSinkWriter::new(
+            self.config.clone(),
+            self.schema.clone(),
+            self.pk_indices.clone(),
+            self.is_append_only,
+        )
+        .await?
+        .into_log_sinker(writer_param.sink_metrics))
+    }
+}
+
+enum SqlOp {
+    Insert(OwnedRow),
+    Merge(OwnedRow),
+    Delete(OwnedRow),
+}
+
+pub struct SqlServerSinkWriter {
+    config: SqlServerConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+    sql_client: SqlClient,
+    ops: Vec<SqlOp>,
+}
+
+impl SqlServerSinkWriter {
+    pub async fn new(
+        config: SqlServerConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        let sql_client = SqlClient::new(&config).await?;
+        let writer = Self {
+            config,
+            schema,
+            pk_indices,
+            is_append_only,
+            sql_client,
+            ops: vec![],
+        };
+        Ok(writer)
+    }
+
+    fn delete_one(&mut self, row: RowRef<'_>) {
+        // TODO: avoid into_owned_row
+        self.ops.push(SqlOp::Delete(row.into_owned_row()));
+    }
+
+    fn upsert_one(&mut self, row: RowRef<'_>) {
+        // TODO: avoid into_owned_row
+        if self.is_append_only {
+            self.ops.push(SqlOp::Insert(row.into_owned_row()));
+        } else {
+            self.ops.push(SqlOp::Merge(row.into_owned_row()));
+        }
+    }
+
+    async fn flush(&mut self) -> Result<()> {
```

Comment:
The semantic of `flush` is to `commit` a stream chunk to downstream. We should use transaction to batch commit ops of a chunk.

Reply:
This `flush` happens whenever this sink writer's buffer is full. Maybe I should rename it?